### PR TITLE
Fixes #8012 where a race condition with useEffect was causing stale renders of values.

### LINKED ISSUE #6699: Reactive variables: Updating nested properties doesn't trigger a re-render
**Intended outcome:**

Following [this](https://www.apollographql.com/blog/local-state-management-with-reactive-variables/) blog post for working with reactive variables, updating a nested value in the reactive variable doesn't re-render the React app.

```ts
import { InMemoryCache, makeVar } from '@apollo/client'
import { uuidv4 } from './utils'

export type Task = {
  id: string
  text: string
  completed: boolean
}

export type Tasks = Task[]

export const cache: InMemoryCache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        tasks: {
          read() {
            return tasks()
          },
        },
      },
    },
  },
})

const initialTasks: Tasks = [
  {
    id: uuidv4(),
    text: 'Task 1',
    completed: true,
  },
  {
    id: uuidv4(),
    text: 'Task 2',
    completed: false,
  },
]

const tasks = makeVar<Tasks>(initialTasks)

export const taskOperations = {
  addtask(text: string) {
    let initialTasks = tasks()
    let task: Task = {
      id: uuidv4(),
      text,
      completed: false,
    }
    tasks([...initialTasks, task])
  },
  toggleTask(id: string) {
    let initialTasks = tasks()
    let index = initialTasks.findIndex(task => task.id === id)
    if (index !== -1) {
      tasks(
        initialTasks.map((task, i) => {
          if (i === index) {
            task.completed = !task.completed
          }
          return task
        })
      )
    }
  },
}
``` 

Here, the `addTask` method triggers a re-render, but the `toggleTask` method doesn't.

**Actual outcome:**

The `toggleTask` method should trigger a re-render

**Versions**

System:
    OS: macOS 10.15.6
  Binaries:
    Node: 12.18.2 - ~/.nvm/versions/node/v12.18.2/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.14.6 - ~/.nvm/versions/node/v12.18.2/bin/npm
  Browsers:
    Chrome: 84.0.4147.89
  npmPackages:
    @apollo/client: ^3.0.2 => 3.0.2

### LINKED ISSUE #8012: When `useReactiveVar` is used in an async callback which awaits a promise, successive updates are not rendered
[Reported by @KonstantinLypskyi](https://github.com/apollographql/apollo-client/issues/7614#issuecomment-823319446).
**Intended outcome:**
The following code should render `1` when the button is pressed.

**Actual outcome:**
`2` is rendered instead.

**How to reproduce the issue:**
https://codesandbox.io/s/use-reactive-var-qobw1?file=/src/App.jsx

**Versions**
TODO

### PR DESCRIPTION
**Pull request by brainkim**
Fixes
#6699 (maybe)
#7694 
#7896
#8012

A race condition with useEffect was causing stale renders of values when reactivevars were called multiple times synchronously.

### CODE REVIEW

**Comment by brainkim** [comment]
Hang on it wasn’t the useLayoutEffect that did it.

File: src/cache/inmemory/reactiveVars.ts [hunk 1]
```diff
@@ -16,19 +16,6 @@ export type ReactiveListener<T> = (value: T) => any;
...
-// iterateObserversSafely for another example of this pattern.
-function consumeAndIterate<T>(set: Set<T>, callback: (item: T) => any) {
-  if (set.size) {
-    const items: T[] = [];
-    set.forEach(item => items.push(item));
-    set.clear();
-    items.forEach(callback);
-  }
-}
-
```
Comment (brainkim) [inline src/cache/inmemory/reactiveVars.ts:31]: I inlined this, but if you’d like me to bring it back lemme know.

File: src/react/hooks/useReactiveVar.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ export function useReactiveVar<T>(rv: ReactiveVar<T>): T {
   // We subscribe to variable updates on initial mount and when the value has
   // changed. This avoids a subtle bug in React.StrictMode where multiple listeners
   // are added, leading to inconsistent updates.
-  useEffect(() => rv.onNextChange(setValue), [value]);
+  useLayoutEffect(() => rv.onNextChange(setValue), [value]);
```
Comment (jcreighton) [inline src/react/hooks/useReactiveVar.ts:12]: What's the reason for useLayoutEffect here?

Comment (brainkim) [reply src/react/hooks/useReactiveVar.ts:12]: So when we call a reactiveVar multiple times synchronously, the `listeners` set in the closure of the reactiveVar is empty, meaning we never get the see a render with the later values. This is because `useEffect` fires an indeterminate amount of time after the `setState` call, and `reactiveVar.onNextChange` is the thing which populates the set for each render. We could alternatively defer the actual calling of the listeners, but this had an odd breaking effect in some of the other tests which test render counts (a practice which I think will cause us pain in the long run but whatever). Using a “layout” effect ensures that each successive call to the reactiveVar is done with listeners set to render the next value.

This is the only change which actually affects behavior; the other change in the cache abstraction was just because I hated to see a weird function with only a single use not inlined while I had the context for it.

TL;DR Slapping a `useLayoutEffect()` seems to make everything work?

File: src/cache/inmemory/reactiveVars.ts [hunk 3]
```diff
@@ -79,7 +66,9 @@ export function makeVar<T>(value: T): ReactiveVar<T> {
           broadcast(cache);
         });
         // Finally, notify any listeners added via rv.onNextChange.
-        consumeAndIterate(listeners, listener => listener(value));
+        const oldListeners = Array.from(listeners);
```
Comment (benjamn) [inline src/cache/inmemory/reactiveVars.ts:69]: For a minute I was worried about the IE compatibility risk of using `Array.from` here, but it turns out [this is not the first place](https://github.com/apollographql/apollo-client/blob/2ec1fe855bc48c50d00ad18482a869fd8f228d78/src/utilities/observables/Concast.ts#L97) where we've used it, and I'm not aware of any complaints from that (though perhaps I should have thought harder about it when I wrote that code).

At this point, it seems like `Array.from` needs polyfilling only in IE11 and earlier: https://caniuse.com/mdn-javascript_builtins_array_from

File: src/react/hooks/useReactiveVar.ts (hunk 2 above)
Comment (benjamn) [reply src/react/hooks/useReactiveVar.ts:12]: @brainkim Is `useLayoutEffect` reliably synchronous, or is the point just that it fires earlier, before the
```ts
Promise.resolve().then(() => {...})
```
callback fires?

I ask because I'm wondering if there might be a way to register the listener synchronously after all, as long as we cancel and re-listen in `useEffect` or `useLayoutEffect`:
```ts
export function useReactiveVar<T>(rv: ReactiveVar<T>): T {
  const value = rv();
  const [, setValue] = useState(value);
  const cancel = rv.onNextChange(setValue);
  useEffect(() => {
    cancel(); // cancel the temporary listener
    return rv.onNextChange(setValue);
  }, [value]);
```
Would this work with `<StrictMode>` double rendering, or would there be two different `cancel` functions and only one would be called?

File: src/cache/inmemory/reactiveVars.ts (hunk 1 above)
Comment (benjamn) [reply src/cache/inmemory/reactiveVars.ts:31]: No, this is good, I love inlining. In my defense, there used to be more callers of `consumeAndIterate` (IIRC).

File: src/react/hooks/__tests__/useReactiveVar.test.tsx [hunk 4]
```diff
@@ -236,5 +236,34 @@ describe("useReactiveVar Hook", () => {
...
+
+    itAsync("works with multiple synchronous calls", async (resolve, reject) => {
+      const counterVar = makeVar(0);
+      function Component() {
+        const count = useReactiveVar(counterVar);
+
+        return (<div>{count}</div>);
+      }
+
+      const { getAllByText } = render(<Component />);
```
Comment (benjamn) [inline src/react/hooks/__tests__/useReactiveVar.test.tsx:248]: Can you try this test in a `<StrictMode>` wrapper (maybe in an additional test, slightly modifying this test), to make sure the [double rendering gotcha](https://stackoverflow.com/questions/61254372/my-react-component-is-rendering-twice-because-of-strict-mode) isn't a problem for this code?

**Review by benjamn — COMMENTED** [review]
I think `useLayoutEffect` is an improvement, but I wonder if we can eliminate the `onNextChange` listening gap completely? Thoughts below.

File: src/react/hooks/__tests__/useReactiveVar.test.tsx (hunk 4 above)
Comment (brainkim) [reply src/react/hooks/__tests__/useReactiveVar.test.tsx:248]: Will do!

File: src/react/hooks/useReactiveVar.ts (hunk 2 above)
Comment (jcreighton) [reply src/react/hooks/useReactiveVar.ts:12]: @benjamn Unfortunately, that will cause the issue you described with `<StrictMode>`.

Comment (brainkim) [reply src/react/hooks/useReactiveVar.ts:12]: As far as I can tell, it’s that’s the `useEffect()` call does not fire between successive var calls, not that `useEffect` fires before or after the promise callback. You can confirm this by changing the callback to a `setTimeout()` probably.

I don’t really have the full context for why we needed to use `useEffect()` in the first place. I guess this PR https://github.com/apollographql/apollo-client/pull/7581 ? My philosophy is that if React is giving me a synchronous escape hatch which doesn’t suffer from timing issues or resource starvation (IN JAVASCRIPT), even if it’s labeled as just being used for “layout,” I’m going to take it.

Now that I’m actually searching it looks like there might be a lot of related issues https://github.com/apollographql/apollo-client/issues/7896 seems like a duplicate. Probably more? Will do a full sweep search.

File: src/cache/inmemory/reactiveVars.ts (hunk 3 above)
Comment (brainkim) [reply src/cache/inmemory/reactiveVars.ts:69]: At the very least, it’s polyfillable, unlike spread operators. BTW Set is not available in <IE10 😛

File: src/react/hooks/__tests__/useReactiveVar.test.tsx (hunk 4 above)
Comment (brainkim) [reply src/react/hooks/__tests__/useReactiveVar.test.tsx:248]: I’m having trouble creating a failing test when swapping the 3.2.6 behavior of useReactiveVar. This was the code example it fixed, but I can’t reproduce it yet in the test suite. I’ll keep looking into it tomorrow. https://codesandbox.io/s/dawn-lake-2yxg3?file=/src/App.js

Comment (brainkim) [reply src/react/hooks/__tests__/useReactiveVar.test.tsx:282]: Done.

**Review by benjamn — APPROVED** [review]
LGTM, thanks for adding that `<StrictMode>` test.
